File: src/content/3.8/code/scala/snippet26.scala
case class StreamF[E, A](h: E, t: A)

implicit def streamFFunctor[E] = new Functor[StreamF[E, ?]] {
  def fmap[A, B](f: A => B)(fa: StreamF[E, A]): StreamF[E, B] =
    ...
}

File: src/content/3.8/code/ocaml/snippet26.ml
(* OCaml is strict by default, so to prevent hanging due
   to the streams being infinite, we make them lazy. *)
type ('e, 'a) stream_f = StreamF of ('e * 'a Lazy.t)

(* Wrapper for parameterized stream *)
module type Type = sig
  type t
end

(* The stream functor could be thus defined as follows: *)
module Stream_Functor (E : Type) : Functor
with type 'a t = (E.t, 'a) stream_f =
struct
  type 'a t = (E.t, 'a) stream_f

  let fmap f (StreamF (e, a)) = StreamF (e, lazy (f (Lazy.force a)))
end

(* And applied to integers *)
module Stream_Functor_Int = Stream_Functor (Int)
